src/S3ClientGUI/mainwindow.cpp

```diff
@@ -356,12 +357,83 @@ void MainWindow::on_LocalContextMenuRequest(const QPoint &point)
     if (info.isFile() && !m_s3model->getRootBucket().isEmpty()) {
         menu.addAction("Upload", this, SLOT(on_upload()));
         menu.exec(m_fsview->mapToGlobal(point));
+    } else if (info.isDir() && index.data().toString() != ".." && !m_s3model->getRootBucket().isEmpty()) {
+        menu.addAction("Upload", this, SLOT(on_uploadDir()));
+        menu.exec(m_fsview->mapToGlobal(point));
     } else {
         return;
     }
 }
 
+QFileInfoList GetFileList(QString path)
+{
+    QDir dir(path);
+    QFileInfoList file_list = dir.entryInfoList(QDir::Files | QDir::Hidden | QDir::NoSymLinks);
+    QFileInfoList folder_list = dir.entryInfoList(QDir::Dirs | QDir::NoDotAndDotDot);
+
+    for (int i = 0; i != folder_list.size(); i++) {
+        QString name = folder_list.at(i).absoluteFilePath();
+        QFileInfoList child_file_list = GetFileList(name);
+        if (child_file_list.empty())
+            file_list.append(name);
+        file_list.append(child_file_list);
+    }
+    if (file_list.empty())
+        file_list.append(path);
+    return file_list;
+}
+
+void MainWindow::on_uploadDir()
+{
+    QFileInfoList fil;
+    QFileInfo info = m_fsview->currentFileInfo();
+    QString parentDirPath = info.absolutePath() + '/';
+    qDebug() << "info=" << qPrintable(info.absolutePath()) << endl;
+    QString remoteBucketName = m_s3model->getRootBucket();
+    QString remotePrefix = m_s3model->getCurrentPrefix();
+    QString KeyName;
+
+    fil = GetFileList(info.absoluteFilePath());
+    for (int i = 0; i < fil.size(); i++) {
+        qDebug() << "fil=" << fil.at(i) << endl;
+    }
+    for (int i = 0; i < fil.size(); ++i) {
+        QString localFile = fil.at(i).absoluteFilePath();
+        QString tmp = localFile;
+        tmp.remove(0, parentDirPath.length() - 1);
+        KeyName = remotePrefix + tmp;
+        qDebug() << "tmp=" << qPrintable(tmp) << endl;
+        qDebug() << "KeyName=" << qPrintable(KeyName) << endl;
+        //if dir is empty,mkdir!
+        if (!fil.at(i).isFile()) {
```

Comment:
What? why create the directory for S3?

Reply:
Maybe someone directory has empty directory.Should I delete it?